ser/train.py

```diff
@@ -0,0 +1,61 @@
+from ser.data import *
+import torch
+from torch import optim
+import torch.nn.functional as F
+
+def train_model(name, epochs, learning_rate, device, train_dataloader, validation_dataloader, model):
+
+    # setup params
+    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
+
+    # record highest validation accuracy
+    highest_accuracy = 0
+    best_epoch = 0
+
+    # train
+    for epoch in range(epochs):
+        for i, (images, labels) in enumerate(train_dataloader):
+            images, labels = images.to(device), labels.to(device)
+            model.train()
+            optimizer.zero_grad()
+            output = model(images)
+            loss = F.nll_loss(output, labels)
+            loss.backward()
+            optimizer.step()
+            print(
+                f"Train Epoch: {epoch} | Batch: {i}/{len(train_dataloader)} "
+                f"| Loss: {loss.item():.4f}"
+            )
+            # validate
+            val_loss = 0
+            correct = 0
+            with torch.no_grad():
```

Comment:
btw, the inefficiency mentioned in the prac is here, usually one would only validate after every epoch (after all the training data has been passed through once) as opposed to within every single batch (just a chunk of the training data). So moving this code block back a step would solve.
You could also consider abstracting away this code block and the above block into separate train batch and val batch functions- as in Claudia's solutions.